### PR TITLE
aggregator retry

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR introduces a retry mechanism for increasing high availability of the aggregator. For now, it only retries on `connection refused` and `no route to host`. The `connection refused` error is of particular interest. Firstly because it can be retried regardless of used HTTP method. Secondly, it turned out it could surface during a graceful shutdown of a downstream server.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 1]
```diff
@@ -38,8 +39,8 @@ func findServicePort(svc *v1.Service, port int32) (*v1.ServicePort, error) {
 	return nil, errors.NewServiceUnavailable(fmt.Sprintf("no service port %d found for service %q", port, svc.Name))
 }
 
-// ResourceLocation returns a URL to which one can send traffic for the specified service.
-func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32) (*url.URL, error) {
+// ResolveEndpoint returns a URL to which one can send traffic for the specified service.
+func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32, seenEndpoints ...*url.URL) (*url.URL, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:43]: doc that the seen endpoints are not returned. One could call it more clearly: avoidEndpoints, excludedEndpoints

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 2]
```diff
@@ -68,6 +69,21 @@ func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.Endpo
 	// Pick a random Subset to start searching from.
 	ssSeed := rand.Intn(len(eps.Subsets))
 
+
+	seenEndpointsToAddressesFn := func(scheme, port string) (sets.String, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:73]: drop Fn postfix.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 3]
```diff
@@ -40,6 +36,11 @@ import (
 	"k8s.io/klog/v2"
 	apiregistrationv1api "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
 	apiregistrationv1apihelper "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/helper"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync/atomic"
+	"time"
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:43]: order

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 4]
```diff
@@ -138,14 +139,51 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		if visitedURL != nil {
+			visitedURLs = append(visitedURLs, visitedURL)
+		}
+
+		if !retryDecorator.RetryIfNeeded() {
+			break
+		}
+		// at this point, we need to retry, reset the counter if there are no remaining endpoints
+		// and let it retry the default number of times
+		if endpointsCount - len(visitedURLs) == 0 {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:167]: endpointsCount == len(visitedURLs)

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 5]
```diff
@@ -138,14 +139,51 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+			visitedURLs = []*url.URL{}
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:168]: this is not intuitive. Why reset it?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 6]
```diff
@@ -138,14 +139,51 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		}
+	}
+
+	// if an error is not retriable and we haven't sent a response to the client return StatusServiceUnavailable
+	if w.(responseWriterInterceptor).StatusCode() == 0 && !w.(responseWriterInterceptor).WasHijacked() {
+		proxyError(w, req, "service unavailable", http.StatusServiceUnavailable)
+	}
+}
+
+func serveHTTP(w http.ResponseWriter, req *http.Request, handlingInfo proxyHandlingInfo, errResponder proxy.ErrorResponder, user user.Info, serviceResolverFn func(namespace, name string, port int32) (*url.URL, error)) *url.URL {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:178]: would rather use `retried bool` as return value.

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:178]: or actually `success bool` ?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 7]
```diff
@@ -225,6 +264,44 @@ func maybeWrapForConnectionUpgrades(restConfig *restclient.Config, rt http.Round
 	return wrappedRT, true, nil
 }
 
+// createRetryDecorator returns a new retry decorator object. It reads the total number of endpoints for the given service from the resolver and use it as the number of reties.
+// In case only a single endpoint will be used it will retry defaultRetryCount.
+func createRetryDecorator(w http.ResponseWriter, errRsp *retriableHijackErrorResponder, handlingInfo proxyHandlingInfo, serviceResolver ServiceResolver, defaultRetryCount int) (*retryDecorator, int, error) {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:269]: why createRetryDecorator and newRetryDecorator? create* is uncommon in Go.

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:269]: newRetryDecoratorForResolver ?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 8]
```diff
@@ -225,6 +264,44 @@ func maybeWrapForConnectionUpgrades(restConfig *restclient.Config, rt http.Round
...
+	}
+
+	if allEPs == 1 {
+		return newRetryDecorator(w.(responseWriterInterceptor), errRsp, true, defaultRetryCount), allEPs, nil
+	}
+
+	return newRetryDecorator(w.(responseWriterInterceptor), errRsp, false, allEPs), allEPs, nil
+}
+
+func (r *proxyHandler) serviceResolverWrapper(visitedEPs []*url.URL) func(namespace, name string, port int32) (*url.URL, error) {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:294]: unclear from the name what it does

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go [hunk 9]
```diff
@@ -28,6 +31,17 @@ type ServiceResolver interface {
 	ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
 }
 
+// ServiceResolverExtended extends ServiceResolver interface by providing a few additional methods for supporting requests retries
+type ServiceResolverExtended interface {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:35]: why not call it RetryServiceResolver?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go [hunk 10]
```diff
@@ -82,3 +96,73 @@ func (r *loopbackResolver) ResolveEndpoint(namespace, name string, port int32) (
...
+	}
+
+	allEndpoints := []*url.URL{}
+
+	for {
+		newEndpoint, err := proxy.ResolveEndpoint(r.services, r.endpoints, namespace, name, port, allEndpoints...)
+		if err != nil {
+			break
+		}
+		allEndpoints = append(allEndpoints, newEndpoint)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:160]: this is kind of odd to probe the ResolveEndpoint method n times to find out that it has n endpoints

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:160]: Would expect something more direct.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+	w.wasHijacked = true
+	return requestHijacker.Hijack()
+}
+
+func (w *responseWriter) WasHijacked() bool {
+	return w.wasHijacked
+}
+
+func (w *responseWriter) StatusCode() int {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go:101]: how about `StatusCode() (hijacked bool, code int)`?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+		if isHTTPVerbRetriable(req) && (knet.IsConnectionReset(err) || knet.IsConnectionRefused(err) || knet.IsProbableEOF(err) ||  isExperimental(err)) {
+			return true
+		}
+		return false
+	}
+	return hr
+}
+
+func isHTTPVerbRetriable(req *http.Request) bool {
+	return req.Method == "GET" || req.Method == "DELETE"
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go:235]: for delete it depends on whether the resource version is given in DeleteOptions afaik.

**Comment by fedebongio** [comment]
/assign @sttts 
Thank you Stefan for taking a look, assigning to you for bug higiene

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go (hunk 12 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go:235]: `DeleteOptions ` doesn't support `ResourceVersion`, I think only GET, LIST and WATCH do.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+type backOff struct {
+	key     string
+	manager *flowcontrol.Backoff
+	retriable
+}
+
+var _ retriable = &backOff{}
+
+func withBackOff(delegate retriable) retriable {
+	return &backOff{"static-key-for-single-host", flowcontrol.NewBackOff(4*time.Second, 30*time.Second), delegate}
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go:169]: for now, hardcoded, let me known if we want to have it configurable.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+		// the error is of particular interest during graceful shutdown of the backend server:
+		// 	net/http2 library automatically retries requests/streams after receiving a GOAWAY frame.
+		// 	this is true for StreamsWithID > LastStreamID (present in the GOAWAY frame) which represents the last stream identifier the server has processed or is aware of.
+		// 	StreamsWithID <= LastStreamID might be fully processed because the server waits until all current streams are done or the timeout expires.
+		// 	in case of the timeout, we will get "http2: server sent GOAWAY and closed the connection" error which is handled by IsProbableEOF() and it is safe to retry only for particular verbs (check isHTTPVerbRetriable method)
+		// 	on the next try a new connection will be opened to the same host and will fail with "connection refused" error because the remote host stopped listening on the port.
+		if knet.IsConnectionRefused(err) {
+			return true
+		}
+		if isHTTPVerbRetriable(req) && (knet.IsConnectionReset(err) || knet.IsProbableEOF(err) || isExperimental(err)) {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go:242]: I understand that some verbs can be retried and I understand why.  However, this particular one cheats the test you're trying to fix.  I can update the test to make sure it doesn't trip this, but I'd rather limit the branches in our proxy overall.

File: cmd/kube-apiserver/app/server.go [hunk 15]
```diff
@@ -694,9 +694,15 @@ func Complete(s *options.ServerRunOptions) (completedServerRunOptions, error) {
 func buildServiceResolver(enabledAggregatorRouting bool, hostname string, informer clientgoinformers.SharedInformerFactory) webhook.ServiceResolver {
 	var serviceResolver webhook.ServiceResolver
 	if enabledAggregatorRouting {
-		serviceResolver = aggregatorapiserver.NewEndpointServiceResolver(
+		// resolve kubernetes.default.svc locally
```
Comment (deads2k) [inline cmd/kube-apiserver/app/server.go:697]: reading top to bottom, but I don't think we require serving on localhost

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go [hunk 16]
```diff
@@ -101,15 +101,7 @@ func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
 	resp, err := rt.RoundTrip(req)
 
 	if err != nil {
-		message := fmt.Sprintf("Error trying to reach service: '%v'", err.Error())
-		resp = &http.Response{
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go:105]: I think this is re-used in several places.  I'm not sure about changing it globally.  Perhaps allowing some kind of option for how to handle this or a different transport entirely?

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 17]
```diff
@@ -38,8 +39,24 @@ func findServicePort(svc *v1.Service, port int32) (*v1.ServicePort, error) {
 	return nil, errors.NewServiceUnavailable(fmt.Sprintf("no service port %d found for service %q", port, svc.Name))
 }
 
-// ResourceLocation returns a URL to which one can send traffic for the specified service.
-func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32) (*url.URL, error) {
+// ResolveEndpoint returns a URL to which one can send traffic for the specified service.
+// This method accepts an optional parameter (excludedEndpoints) for specifying a list of endpoints that must be excluded.
+func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32, excludedEndpoints ...*url.URL) (*url.URL, error) {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:44]: I'm surprised that exclusion is related to retries.  A second PR hiding in here for a different purpose?

**Pull request closed without merging**